### ISSUE TITLE
Server to client TextDocumentItem/

### ISSUE THREAD

**Issue by ratmice** [microsoft/language-server-protocol#1458]
I was curious if there were any plans for a server to client mechanism for sending `TextDocumentItem`'s from server to client,
currently which I understand is only used for sending client documents to the server.

As well as a mechanism for specifying a uri for each active lsp by which a client can recognize documents from the server inside `TextDocumentIdentifiers`. And a new request by which clients could retrieve a server document.

The LSP that I am working generates some extremely long diagnostics, unfortunately diagnostics are typically displayed by editors in a fashion which is only suitable for extremely short messages.  In these cases of excessively long diagnostics, it would be nice if we could then refer to server generated documents via their URL in via the `relatedDocuments` component of the diagnostic.

**Comment by rwols** [comment]
> The LSP that I am working generates some extremely long diagnostics, unfortunately diagnostics are typically displayed by editors in a fashion which is only suitable for extremely short messages.

Have you considered shortening the messages, and then referring to "more in-depth" via [CodeDescription](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeDescription)?

**Comment by ratmice** [comment]
> > The LSP that I am working generates some extremely long diagnostics, unfortunately diagnostics are typically displayed by editors in a fashion which is only suitable for extremely short messages.
> 
> Have you considered shortening the messages, and then referring to "more in-depth" via [CodeDescription](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#codeDescription)?

It not static text, but entirely generated with information derived from the source text.
If you are familiar with parsers, yacc and the like it essentially contains a representation of the parsing stack which arises from a conflict in the grammar (where my lsp is dealing with yacc files.).

**Comment by robbycbennett** [comment]
The `rust-analyzer` extension has a workaround for dynamically-generated `codeDescription` fields. [In these parts of their code](https://github.com/search?q=repo%3Arust-lang%2Frust-analyzer%20URI_SCHEME&type=code) they create a [TextDocumentContentProvider](https://code.visualstudio.com/api/references/vscode-api#TextDocumentContentProvider).

However, this is a fix for just VS Code and rust-analyzer, so it's not an ideal solution for the LSP as a whole.

**Comment by RobbyCBennett** [comment]
For a long description, the LSP should support both dynamic and static descriptions. Here's an idea of how this could be implemented.

In `CodeDescription` add `dynamicDescription` as an optional string. Make `href` optional.
```ts
export interface CodeDescription {
	/**
	 * An URI to open with general information about the diagnostic error (e.g.
	 * paragraphs of a description, code examples, exceptions, links to sources).
	 * The URI can start with "file:", "https:", "diagnostic:", or other.
	 */
	href?: URI;

	/**
	 * A string with unique information about this particular instance of the
	 * diagnostic error (e.g. a warning that mentions your variable name
	 * specifcally and the function it was defined in).
	 */
	dynamicDescription?: string;
}
```

Add a `textDocument/publishDiagnosticLongDescription` method, as a notification from the server to the client. This is to avoid repetition of the long descriptions.

```ts
interface DiagnosticLongDescriptionParams extends WorkDoneProgressParams {
	/**
	 * An array of long descriptions for each diagnostic
	 */
	diagnosticLongDescriptions: DiagnosticLongDescription[];
}
```

```ts
interface DiagnosticLongDescription {
	/**
	 * The diagnostic's code, which might appear in the user interface.
	 */
	code: integer | string;

	/**
	 * The static long description of a diagnostic (e.g. paragraphs of a
	 * description, code examples, exceptions, links to sources).
	 */
	longDescription: string;
}
```

**Comment by ratmice** [comment]
@RobbyCBennett Yeah, implementing a `TextDocumentContentProvider` is exactly what I ended up doing as well eventually.

**Comment by ratmice** [comment]
@RobbyCBennett I believe your second comment was actually intended for #1461

**Comment by robbycbennett** [comment]
Okay. I'll move it to there.
